perf(linter/no-unused-vars): do not construct `Regex` for default ignore pattern

camchenry opened the pull request:
Profiling shows that constructing the `Regex` for `no-unused-vars` takes up a good chunk of time for small files. Since the default pattern for ignored variables is just `^_`, we can more simply implement this with `.starts_with('_')`.

<img width="1439" alt="image" src="https://github.com/user-attachments/assets/7a4abb8c-2c0e-4252-8df4-e5b1880e8209">


I decided to create a new `Option`-like enum that has three states: `Some` and `None` which are like the typical `Option` states, and another `Default` state which is a special case. The `Default` state represents that matching code should use the default pattern (if applicable). I decided not to keep using `Option`, because I didn't want to overload the `None` state by confusing whether or not it was the default state, or the user explicitly overrided it as `None`.

The theory here is that most users are not going to pass a custom ignore pattern, so we can avoid needing to create a `Regex` in most cases, which greatly speeds up the initialization for this rule.

camchenry commented:
* **#6590** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6590?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @camchenry and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

camchenry commented:
codspeed shows +0% but it's closer to ~1% locally:

```
hyperfine --warmup 100 --shell=none --runs 9000 './oxlint-main --silent ./RadixUIAdoptionSection.jsx' './oxlint-unused-regex --silent ./RadixUIAdoptionSection.jsx'
Benchmark 1: ./oxlint-main --silent ./RadixUIAdoptionSection.jsx
  Time (mean ± σ):       3.4 ms ±   0.1 ms    [User: 1.1 ms, System: 1.3 ms]
  Range (min … max):     3.2 ms …   3.9 ms    9000 runs

Benchmark 2: ./oxlint-unused-regex --silent ./RadixUIAdoptionSection.jsx
  Time (mean ± σ):       3.4 ms ±   0.1 ms    [User: 1.1 ms, System: 1.3 ms]
  Range (min … max):     3.0 ms …   4.2 ms    9000 runs

Summary
  ./oxlint-unused-regex --silent ./RadixUIAdoptionSection.jsx ran
    1.01 ± 0.05 times faster than ./oxlint-main --silent ./RadixUIAdoptionSection.jsx
```

DonIsaac commented on crates/oxc_linter/src/rules/eslint/no_unused_vars/options.rs:
```diff
@@ -207,18 +207,92 @@ pub struct NoUnusedVarsOptions {
+/// which represents the default ignore pattern for when no pattern is
+/// explicitly provided.
+#[derive(Debug, Clone, Copy)]
+pub enum IgnorePattern<R> {
```
Do we need the type parameter?

camchenry committed:
perf(linter/no-unused-vars): do not construct `Regex` for default ignore pattern (#6590)

Profiling shows that constructing the `Regex` for `no-unused-vars` takes up a good chunk of time for small files. Since the default pattern for ignored variables is just `^_`, we can more simply implement this with `.starts_with('_')`.

<img width="1439" alt="image" src="https://github.com/user-attachments/assets/7a4abb8c-2c0e-4252-8df4-e5b1880e8209">

I decided to create a new `Option`-like enum that has three states: `Some` and `None` which are like the typical `Option` states, and another `Default` state which is a special case. The `Default` state represents that matching code should use the default pattern (if applicable). I decided not to keep using `Option`, because I didn't want to overload the `None` state by confusing whether or not it was the default state, or the user explicitly overrided it as `None`.

The theory here is that most users are not going to pass a custom ignore pattern, so we can avoid needing to create a `Regex` in most cases, which greatly speeds up the initialization for this rule.
